containers-tests/tests/tree-properties.hs

```diff
@@ -150,6 +176,12 @@ prop_foldMap t =
   foldMap (:[]) t === toList t .&&.
   foldMap (:[]) t === foldMapDefault (:[]) t
 
+-- We use UnitalMagma with foldMap to test that the structure of the fold
+-- follows that of the tree. This is a desirable property to have.
```

Comment:
It's not *always* a desirable property, but it's a desirable property for `Tree`. I would love to be able to articulate exactly why, but I'm struggling. I'll give you a few examples of yes/no, and maybe you can figure out a concise way to say what needs to be said?

* `Data.Tree`: the structure is lazy and exposed concretely, with no external (or fancy-type-induced) data structure invariants—the fold should follow the structure.
* `Data.Set`: the structure is strict and has a size annotation invariant we can use easily to avoid unnecessarily appending `mempty`s at the leaves—the fold should not precisely follow the structure.
* A scheduled banker's queue: A lazy fold is expected to require $O(1)$ actual time per element; following the structure would lead to an $O(n)$ pause to reverse the rear, so the fold must not follow the structure. (A strict fold is technically allowed to follow the structure, but doing so might be bad for caching or concurrency and should probably be avoided.)

Perhaps it would be best to avoid generalities, and just try to explain why following the structure of the tree is appropriate for this structure?

Reply:
Too strict.

```haskell
foldMap (:[]) (Node x undefined) = x : _|_
foldMap1 (:[]) (Node x undefined) = _|_
```

and I don't think the latter can be avoided.

Reply:
Ah those pesky bottoms.

> Perhaps it would be best to avoid generalities, and just try to explain why following the structure of the tree is appropriate for this structure?

Do you have an example of a `Monoid` where this implementation is preferable? I'm finding it hard to come up with one. For `Semigroup` we have `Last`.

Reply:
Well, `First a` is a simple example, or `[a]` as I demonstrated already. I'm not sure what you're asking.

Reply:
I mean a `Monoid` where `foldMap` according to the tree structure is preferable to the default that uses `foldr`.

Reply:
Well, there's `Maybe (Last a)`, AKA `Data.Monoid.Last a`.

Reply:
I guess that works. Updated, how does it look?